src/include/function/gds/gds.h

```diff
@@ -1,129 +1,80 @@
     }
+    std::shared_ptr<processor::FactorizedTable> getResultTable() const { return resultTable; }
 
-    virtual std::unique_ptr<GDSBindData> copy() const {
+    std::unique_ptr<TableFuncBindData> copy() const override {
         return std::make_unique<GDSBindData>(*this);
     }
 
-    template<class TARGET>
-    TARGET* ptrCast() {
-        return common::ku_dynamic_cast<TARGET*>(this);
-    }
+private:
+    std::shared_ptr<processor::FactorizedTable> resultTable;
 };
 
-// Base class for every graph data science algorithm.
-class KUZU_API GDSAlgorithm {
-protected:
-    static constexpr char NODE_COLUMN_NAME[] = "node";
+struct KUZU_API GDSFuncSharedState : public TableFuncSharedState {
+    std::unique_ptr<graph::Graph> graph;
 
-public:
-    GDSAlgorithm() = default;
+    GDSFuncSharedState(std::shared_ptr<processor::FactorizedTable> fTable,
+        std::unique_ptr<graph::Graph> graph)
+        : TableFuncSharedState{}, graph{std::move(graph)}, factorizedTablePool{std::move(fTable)} {}
 
-    GDSAlgorithm(const GDSAlgorithm& other) {
-        if (other.bindData != nullptr) {
-            bindData = other.bindData->copy();
-        }
-        sharedState = other.sharedState;
-    }
-
-    virtual ~GDSAlgorithm() = default;
-
-    virtual std::vector<common::LogicalTypeID> getParameterTypeIDs() const { return {}; }
-
-    virtual binder::expression_vector getResultColumns(
-        const function::GDSBindInput& bindInput) const = 0;
-
-    virtual void bind(const GDSBindInput& input, main::ClientContext& context) = 0;
-
-    const GDSBindData* getBindData() const { return bindData.get(); }
-
-    // Note: The reason this field is set separately here and not inside constructor is that the
-    // original GDSAlgorithm is constructed in the binding stage. In contrast, sharedState is
-    // constructed during mapping phase. Also as a coding pattern, "shared state objects" used
-    // inside physical operators are set inside the init functions inside the operators. Although
-    // GDSAlgorithm is not an operator, it is close to an operator. It is simply used by the GDSCall
-    // operator, which is basically a wrapper around a GDSAlgorithm.
-    void setSharedState(std::shared_ptr<processor::GDSCallSharedState> _sharedState) {
-        sharedState = _sharedState;
-    }
+    void setGraphNodeMask(std::unique_ptr<common::NodeOffsetMaskMap> maskMap);
+    common::NodeOffsetMaskMap* getGraphNodeMaskMap() const { return graphNodeMask.get(); }
 
-    virtual void exec(processor::ExecutionContext* executionContext) = 0;
+public:
+    processor::FactorizedTablePool factorizedTablePool;
 
-    virtual std::unique_ptr<GDSAlgorithm> copy() const = 0;
+private:
+    std::unique_ptr<common::NodeOffsetMaskMap> graphNodeMask = nullptr;
+};
 
-    template<class TARGET>
-    TARGET& cast() {
-        return common::ku_dynamic_cast<TARGET&>(*this);
-    }
+// Base class for every graph data science algorithm.
+class KUZU_API GDSAlgorithm {
```

Comment:
I think it is a pure utility class right now, should we rename it to `GDSFunction`?
Because we have `GDSFuncBindData` and `GDSFuncSharedState` and i don't see we use the term `GDSAlgo` in other places.